src/t_stream.c

```diff
@@ -509,16 +513,31 @@ int streamAppendItem(stream *s, robj **argv, int64_t numfields, streamID *added_
             lp = NULL;
         } else if (server.stream_node_max_entries) {
             int64_t count = lpGetInteger(lpFirst(lp));
-            if (count >= server.stream_node_max_entries) lp = NULL;
+            if (count >= server.stream_node_max_entries) {
+                /* Shrink extra pre-allocated memory */
+                lp = lpShrinkToFit(lp);
+                if (ri.data != lp)
+                    raxInsert(s->rax,ri.key,ri.key_len,lp,NULL);
+                lp = NULL;
+            }
         }
     }
 
     int flags = STREAM_ITEM_FLAG_NONE;
     if (lp == NULL) {
         master_id = id;
         streamEncodeID(rax_key,&id);
-        /* Create the listpack having the master entry ID and fields. */
-        lp = lpNew();
+        /* Create the listpack having the master entry ID and fields.
+         * Pre-allocate some bytes when creating listpack to avoid realloc on
+         * every XADD. Since listpack.c uses malloc_size, it'll grow in steps,
+         * and won't realloc on every XADD.
+         * When listpack reaches max number of entries, we'll shrink the
+         * allocation to fit the data. */
+        size_t prealloc = STREAM_LISTPACK_MAX_PRE_ALLOCATE;
+        if (server.stream_node_max_bytes > 0 && server.stream_node_max_bytes < prealloc) {
```

Comment:
can stream_node_max_bytes be 0? if so, we should fix it..
also, i'd like to remind that we have a MIN macro

Reply:
I think `0` means unlimited here.

Reply:
i see that it is possible for it to be 0 (no idea why) but in this case i think prealloc should be 0... it's a bit weird that:
stream_node_max_bytes == 2 -> prealloc = 2
stream_node_max_bytes == 1 -> prealloc = 1
stream_node_max_bytes == 0 -> prealloc = 4096

Reply:
```c
if (lp == NULL || lp_bytes >= server.stream_node_max_bytes) {
        master_id = id;
        streamEncodeID(rax_key,&id);
        /* Create the listpack having the master entry ID and fields. */
        lp = lpNew();
        lp = lpAppendInteger(lp,1); /* One item, the one we are adding. */
        lp = lpAppendInteger(lp,0); /* Zero deleted so far. */
```
seems like it'll create a new lp on every streamAppendItem... unless i'm missing something

Reply:
According to this code: https://github.com/redis/redis/blob/unstable/src/t_stream.c#L509
I think `0` is more like `+inf`?

Reply:
> seems like it'll create a new lp on every streamAppendItem... unless i'm missing something

`lp` is set to `NULL` only when a new one needs to be created.

Reply:
oh ok i think i just didn't look in `unstable`... ok then, +1

Reply:
Seems like it can be zero. The `0` (before LONG_MAX) in the code from config.c below is the minimum:

```C
    createSizeTConfig("stream-node-max-bytes", NULL, MODIFIABLE_CONFIG, 0, LONG_MAX, server.stream_node_max_bytes, 4096, MEMORY_CONFIG, NULL, NULL),
```

The only place this config is used is in t_stream.c:

```C
    /* First of all, check if we can append to the current macro node or
     * if we need to switch to the next one. 'lp' will be set to NULL if
     * the current node is full. */
    if (lp != NULL) {
        if (server.stream_node_max_bytes &&
            lp_bytes >= server.stream_node_max_bytes)
        {
            lp = NULL;
```

... so zero means "no limit" here.